expect: Highlight substring differences when matcher fails, part 1

pedrottimark opened the pull request:
## Summary

Fixes #6881 although space and no-break space still look the same

If positive assertion fails with **one-line** strings for both expected and received:

Then if character difference with “semantic cleanup” has **any common** substring:

Display changes similar to Changed Files on GitHub:

* Expected: inverse green for deleted characters
* Received: inverse red for inserted characters

Else if **no common** substring, display without any inverse colors

**Build**:

* `packages/expect/build-es5/index.js` is about 1M bytes, increased < 40K bytes
* although I added `diff-sequences` dependency, it was already in `build-es5` bundle
* added to `expect/tsconfig.json` see https://github.com/facebook/jest/pull/8448#issuecomment-491610188
* ~~review of `packages/expect/src/cleanupSemantic.d.ts` is especially welcome~~
* ~~see https://github.com/facebook/jest/pull/8448#issuecomment-491050415 about `packages/expect/build/cleanupSemantic.d.js`~~

**License**:

Does anyone know if I have correctly handled code with Apache License, Version 2.0?

* added comment describing changes to `cleanupSemantic.js`
* added file to `.eslintignore` and `.prettierignore`
* added file path as exception to `scripts/checkCopyrightHeaders.js`

**Residue**:

* Decide how to display results for **multi-line** strings (especially deleted or inserted newlines)
* Discover whether long strings cause **performance** risk as had been for many lines
* Apply to more changed strings (for example, property values) in future data-driven diff
* Adapt for `toThrow(object)` which does not call the same helper function
* Explore how to display results of failed **substring match** for `toContain`, `toMatch`, or `toThrow` [EDIT: and found its effectiveness to be disappointingly hit or miss]

## Test plan

Existing tests pass, because they have no common substring with “semantic cleanup”

Added 1 snapshot each:

* `toBe` and then added a second test for empty string, see https://github.com/facebook/jest/pull/8448#issuecomment-491844721
* `toEqual`
* `toHaveProperty(path, value)`
* `toStrictEqual`

See also pictures in following comment

pedrottimark commented:
Example pictures baseline at left and **improved at right**

Change a word:

<img width="1000" alt="toBe" src="https://user-images.githubusercontent.com/11862657/57483586-2a6c6e00-7275-11e9-970e-babfe12a7945.png">

Change spaces to no-break spaces:

<img width="1000" alt="toEqual" src="https://user-images.githubusercontent.com/11862657/57483590-2d675e80-7275-11e9-887e-d5b035839c21.png">

Edit around a phrase:

<img width="1000" alt="toHaveProperty" src="https://user-images.githubusercontent.com/11862657/57483603-39532080-7275-11e9-9527-f97c6568b2db.png">

I added a simpler example (than the above) to unit tests

Edit around and in a phrase:

<img width="1000" alt="toStrictEqual" src="https://user-images.githubusercontent.com/11862657/57483663-54259500-7275-11e9-89eb-eb2de80f5789.png">

pedrottimark commented:
I am not sure if the `.d.ts` file in `src` directory is correctly included in the build:

Here is `packages/expect/build/cleanupSemantic.d.js`

```js
'use strict';

function _defineProperty(obj, key, value) {
  if (key in obj) {
    Object.defineProperty(obj, key, {
      value: value,
      enumerable: true,
      configurable: true,
      writable: true
    });
  } else {
    obj[key] = value;
  }
  return obj;
}
```

pedrottimark commented:
Thinking ahead to remove limitation on one-line strings in future pull request, what do y’all think about displaying inserted or deleted line break as U+21B5 downwards arrow with corner leftwards:

<img width="1000" alt="multiline" src="https://user-images.githubusercontent.com/11862657/57562004-5531f180-735d-11e9-9f7e-271859960038.png">

SimenB commented:
Oh my god, I'm so excited! 😀 

> I am not sure if the `.d.ts` file in `src` directory is correctly included in the build:

It should be, we had the same thing `@jest/reporters` and it worked

SimenB commented on CHANGELOG.md:
```diff
@@ -2,6 +2,8 @@
 
 ### Features
 
+- `[expect]` Improve report when matcher fails, part 20 ([#8448](https://github.com/facebook/jest/pull/8448))
```
substring highlighting deserves explicit mention 😀

SimenB commented:
Would it make sense for the vendored file to live in `diff-sequences` and not `expect`?

SimenB commented:
And yes, rendering some explicit newline character would be awesome, like ava does

pedrottimark commented:
@SimenB it’s related to your https://github.com/facebook/jest/issues/8085#issuecomment-470924242 although `jest-reporters` had no `src/*.js` file:

> It seems like those files are not copied over by `tsc`

> there are some `d.js` files there, funnily enough

I think the problem is https://github.com/facebook/jest/blob/master/scripts/build.js#L122-L125

* it does **not** copy the `.d.ts` file because it does match `TS_FILES_PATTERN`
* but then I assume that babel compiles it as if it is source code

Does it make sense to swap the logic to avoid double negative in `if-else` as follows:

```js
if (
  micromatch.isMatch(file, JS_FILES_PATTERN) ||
  (micromatch.isMatch(file, TS_FILES_PATTERN) &&
    !micromatch.isMatch(D_TS_FILES_PATTERN))
) {
  // compile file
} else {
  // copy file
}
```

SimenB commented:
They should work - we deleted some in https://github.com/facebook/jest/pull/8294 which worked fine previously.

It's the tsc run that should copy it, not babel. Try to run `build:ts`

pedrottimark commented:
Good question about packaging. `jest-matcher-utils` is a possibility after interface becomes clear

The reason to keep `semanticCleanup` out of `diff-sequences` is same reason we did not just depend on the excellent `diff-match-patch` package:

* it has features we don’t need that I’m not sure tree shaking could get rid of
* its class interface it is harder to adapt than our callback interface

pedrottimark commented:
@SimenB this is making me learn more about Jest build scripts and TypeScript config :)

`yarn build-clean` and then `yarn build:ts` creates `.d.ts` and `.d.ts.map` only for `src/*.ts` files

* It does not copy through `src/*.d.ts` to `build` directory
* Even, by the way, if I add it explicitly to a `files` array in `expect/tsconfig.json` file

Compiling as `.ts` there are 22 errors but `allowJs: true` and `// @ts-check` comment only 7 errors because `tsc` uses JSDoc comments. However that doesn’t give us the `.d.ts` file (`Option 'allowJs' cannot be specified with option 'declaration'` error) which is the problem to solve.

pedrottimark commented:
Solved by renaming as `cleanupSemantic.ts` which was not as hard as I had supposed:

* Converted `Diff` to class from constructor and prototype
* Added type annotations for arguments and return values

Finding that tricky `Diff.prototype.length = 2;` isn’t possible in TypeScript took most of the time ;)

**Question**: in `packages/expect/tsconfig.json` is it an optional or necessary manual edit to add `{"path": "../diff-sequences"},` to `references` for new dependency?

jeysal commented:
> **Question**: in `packages/expect/tsconfig.json` is it an optional or necessary manual edit to add `{"path": "../diff-sequences"},` to `"references"` array for new dependency?

Mandatory (TSC won't ever look at the folder if it's not referenced) - if it works without, it's probably "lucky" because it's also referenced from ´jest-diff`

SimenB commented:
Ah, the way we solved it for Istanbul was a `/// reference` entry. Since you've converted already that's probably a better solution :)

pedrottimark commented:
While writing tests for multi-line strings, I found an edge case of comparison to empty string

Highlighting the non-empty string at left seems less helpful than no highlight at right:

<img width="1000" alt="empty one-line expected" src="https://user-images.githubusercontent.com/11862657/57629136-46b82580-7569-11e9-91d0-cea5b7571f09.png">

<img width="1000" alt="empty one-line received" src="https://user-images.githubusercontent.com/11862657/57629152-4ddf3380-7569-11e9-862f-67d11c367371.png">

Added a test and then updated it after adding the missing condition

For the test where I found the problem, do you prefer the report at left or right, and why?

<img width="1000" alt="empty multi-line received" src="https://user-images.githubusercontent.com/11862657/57629117-3ef88100-7569-11e9-939c-e3054535d931.png">

SimenB commented:
I prefer the right one, since it's explicit about the value of "Received". The left one looks like a bug due to it missing received

pedrottimark commented:
* moved `printDiffOrStringify` from `expect` to `jest-matcher-utils` package
* updated one snapshot test because it is incorrect to display diff for asymmetric matcher
* part 2: specialized diff with highlight for multiline strings and also use in `toThrow(object)`
* part 3: when received is a string, call `printDiffOrStringify` for report when snapshot fails
